Update distribution.py to check for integer shape

Anjum48 opened the pull request:
Added additional check to see if the Distribution shape only contains integers. Raises TypeError.

See https://github.com/pymc-devs/pymc3/issues/836

Anjum48 committed:
Update distribution.py

Added check to see if the Distribution shape only contains integers. Raises TypeError.

Anjum48 committed:
Update distribution.py

Forgot a to add " after int32 :(

Anjum48 committed:
Update distribution.py

Cast shape to a string so that TypeError can be printed

Anjum48 committed:
Update distribution.py

Attempt to fix the "TypeError (Can't convert 'tuple' object to str implicitly)" in line 40

Anjum48 committed:
Update distribution.py

Another attempt to fix the TypeError printout

Anjum48 committed:
Update distribution.py

Gave up trying to include the inputted shape in the TypeError print out - i'm not sure if my basic Python skills can stretch that far!

Anjum48 committed:
Update distribution.py

I think that empty lists/tuples are causing the TypeError to be raised, causing the build to fail.  Added a len(self.shape) > 0 constraint

fonnesbeck commented:
Would it be appropriate to simply cast to integers? Could do a modulus test if its a float so that we don't silently remove fractional values.

jsalvatier commented:
I'm slightly concerned that we would want to allow variable length arrays for some reason, but it hasn't come up yet.

kyleam commented:
The proposed check would miss `np.int64`.  `np.issubdtype` is an another way to handle this  (though I didn't check whether it was added after minimum supported numpy version).  However, I  don't see much advantage in adding this check since the error theano signals already has a pretty clear message.

Anjum48 commented:
I agree that the Theano message does state that the shape has an issue, but the traceback it didn't point in a clear manner where the issue might lie (especially for someone who has no experience with Theano). I didn't know about the `np.subtype` trick which is quite neat. I've changed it, but if you think there's no advantage, we can drop it

Anjum48 committed:
Used np.issubdtype to check for integers instead

Anjum48 committed:
Changed integer checking to use a modulus type approach

The np.issubdtype method doesn't like shapes like `32.0`. Using np.floor and looking for any differences when compared to the original shape should allow these through

twiecki commented:
Squashed and merged with https://github.com/pymc-devs/pymc3/commit/f97cf829f7a78465f8c2af267f7030ae3e51115d. Thanks!
